Cabal/Distribution/Version.hs

```diff
@@ -169,24 +169,55 @@ instance Text Version where
 mkVersion :: [Int] -> Version
 -- TODO: add validity check; disallow 'mkVersion []' (we have
 -- 'nullVersion' for that)
-mkVersion ns = case ns of
-    [] -> nullVersion
-    [v1]          | v1 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) 0 0 0)
-    [v1,v2]       | v1 <= 0xfffe, v2 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) (v2+1) 0 0)
-    [v1,v2,v3]    | v1 <= 0xfffe, v2 <= 0xfffe, v3 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) (v2+1) (v3+1)  0)
-    [v1,v2,v3,v4] | v1 <= 0xfffe, v2 <= 0xfffe, v3 <= 0xfffe, v4 <= 0xfffe
-      -> PV0 (mkW64 (v1+1) (v2+1) (v3+1) (v4+1))
-    v1:vs -> PV1 v1 vs
+mkVersion []                    = nullVersion
+mkVersion (v1:[])
+  | inWord16VerRep1 v1          = PV0 (mkWord64VerRep1 v1)
```

Comment:
Negative values are considered invalid input to `mkVersion` and hence undefined behaviour; consequently the tests made sure to test only with non-negative `[Int]`s...